### PR TITLE
Overlord v1

### PR DESCRIPTION
**Pull request by Manbearpixel**
Added BDD testing files. Gave overlord a Front-End look using HAML and AngularJS. Connected front with back for timer and bomb interaction.

### CODE REVIEW

File: overlord/features/step_definitions/bomb_steps.rb [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  fill_in_deactive_code(code)
+end
+
+When(/^the villian clicks "([^"]*)"$/) do |button_text|
+  click_button(button_text)
+end
+
+When(/^the bomb has been deployed$/) do
+  click_button("DEPLOY")
+  using_wait_time 10 do
```
Comment (johnmcconnell) [inline overlord/features/step_definitions/bomb_steps.rb]: http://www.rubydoc.info/github/jnicklas/capybara/Capybara.using_wait_time booo....
These function cause unit tests to be non-deterministic. I get that it is running Javascript, does your driver support javascript eval?

```ruby
result = driver.javascript_exec('manipulate something here and wait for execution')
```

Comment (johnmcconnell) [inline overlord/features/step_definitions/bomb_steps.rb]: Maybe something like:
```ruby
driver.execute_script(
  "$('#activateBomb > button').first().click()"
)
```

**Comment by johnmcconnell** [comment]
Unsure why you have haml and erb. otherwise :+1:

File: overlord/views/index.haml [hunk 2]
```diff
@@ -0,0 +1,63 @@
+%html{ "data-ng-app" => "EvilBombApp", "data-ng-controller" => "HomeCtrl" }
```
Comment (johnmcconnell) [inline overlord/views/index.haml]: Why both HAML and ERB?

Comment (Manbearpixel) [inline overlord/views/index.haml]: Ah, forgot to take that out. Was originally going to go with ERB but changed my mind and decided to try out HAML.

File: overlord/features/step_definitions/bomb_steps.rb (hunk 1 above)
Comment (Manbearpixel) [inline overlord/features/step_definitions/bomb_steps.rb]: I'll look into this! When I was pairing with someone who assisted with basic Cucumber, he had advised `using_wait_time` to ensure there was a delay to capture JS results.

Comment (johnmcconnell) [inline overlord/features/step_definitions/bomb_steps.rb]: Execute script blocks until it is finished (I'm 99% sure). This eliminates guessing the "correct time". Accordingly, and a wait time creates a undetermined part of the test that can cause it to randomly fail!

File: overlord/features/bomb.feature [hunk 3]
```diff
@@ -0,0 +1,39 @@
+Feature: EVIL Bomb
+  As a villian
+  I want to have an EVIL bomb
+  So that I can be a Supervillian
+
+  Background:
+    Given the villian is on the homepage
+    When the villian enters the bomb activation code "1233"
```
Comment (paulghaddad) [inline overlord/features/bomb.feature]: You can abstract steps like this:

'When the villian enters a correct bomb activation code'

File: overlord/features/bomb.feature [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    And the villian enters the bomb deactivation code "3872"
+
+  Scenario: Deploy Bomb
+    When the villian clicks "DEPLOY"
```
Comment (paulghaddad) [inline overlord/features/bomb.feature]: Try to write your step definitions in more higher level terms, such as:

`When the bomb is deployed`

File: overlord/features/bomb.feature [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    Then the villian expects the bomb status to be "ready for activation"
```
Comment (paulghaddad) [inline overlord/features/bomb.feature]: `Then the bomb should be ready for activation`

File: overlord/features/bomb.feature [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  Scenario: Activate Bomb
+    When the bomb has been deployed
+    And the villian enters the bomb activation code "1233"
+    And the villian clicks "ACTIVATE"
+    Then the villian expects the bomb status to be "activated"
+
+  Scenario: Deactivate Bomb
+    When the bomb has been deployed
+    And the bomb has been activated
+    And the villian enters the bomb deactivation code "3872"
```
Comment (paulghaddad) [inline overlord/features/bomb.feature]: Try to write your feature tests at a high level of abstraction. You should be able to combine lines 22-23 into one step definition, such as `When the bomb has been activated` (encompassing both deployment and activation steps). Then lines 24-25 can be combined in a similar fashion.

File: overlord/features/bomb.feature [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    And the villian clicks "DEACTIVATE"
+    Then the villian expects the bomb status to be "ready for activation"
+
+  Scenario: Enter Incorrect Deactivation Code 3 Times
+    When the bomb has been deployed
+    And the bomb has been activated
+    And the hero enters an incorrect deactivation code three times
```
Comment (paulghaddad) [inline overlord/features/bomb.feature]: Line 31 is an excellent example of writing higher level step definitions. You could have written `And the hero enters an incorrect deactivation code` three times. But your step definition is much more readable and concise.

**Comment by paulghaddad** [comment]
@Manbearpixel Overall, you did a very nice job on this. The one thing to focus on, in addition to John's comments, is your feature tests. Take a look at my comments, and refactor, focusing on writing the tests at a higher level of abstraction, almost like you're talking to a person.

**Pull request closed without merging**